aips/on-chain-randomness-impl.md

```diff
@@ -0,0 +1,419 @@
+
+See reference implementation [here](https://github.com/aptos-labs/aptos-core/blob/1de391c3589cf2d07cb423a73c4a2a6caa299ebf/consensus/src/state_computer.rs#L203).
+
+#### Move VM changes
+
+The Move VM needs to be able to execute the new transaction variants `BlockMetadataExt` and `ValidatorTransaction::DKGResult`,
+which are the 2 steps of an async reconfiguration.
+
+A `BlockMetadataExt` transaction should do what a `BlockMetadata` transaction does plus the following differences.
+
+- On current epoch timeout, trigger async reconfiguration instead of instant reconfiguration.
+  Recall that to trigger async reconfiguration means to:
+  - finalize the next validator set,
+  - lock the validator set resources,
+  - emit a `DKGStartEvent` that contains the next validator set.
+- Write the block randomness seed on chain, if it is available.
+
+Like `BlockMetadata` transactions, `BlockMetadataExt` transactions should never abort.
+See the reference implementation [here](https://github.com/aptos-labs/aptos-core/blob/1de391c3589cf2d07cb423a73c4a2a6caa299ebf/aptos-move/aptos-vm/src/aptos_vm.rs#L1935).
+
+A `ValidatorTransaction::DKGResult` transaction should do the following.
+
+- (a) Verify the aggregated DKG transcript included.
+- Unlock the validator set resources.
+- Apply all pending on-chain config changes.
+- Publish DKG transcript on chain.
+- Whatever is done in `BlockMetadata` transaction when epoch timeout is detected.
+
+Malicious validators can cause step (a) to fail, in which case should cause the transaction to be discarded.
+The remaining steps should never abort.
+See the reference implementation [here](https://github.com/aptos-labs/aptos-core/blob/1de391c3589cf2d07cb423a73c4a2a6caa299ebf/aptos-move/aptos-vm/src/validator_txns/dkg.rs#L70).
+
+
+## Testing (Optional)
+
+Unit tests, smoke tests, forge tests.
+
+Deploy a test network called `randomnet` which lasts for several months.
+
+Further stress / performance testing in `previewnet`.
+
+## Risks and Drawbacks
+
+### Disaster Scenarios
+
+**DKG loses liveness**. In case of implementation bugs/bad configurations, DKG may never finish, so the epoch change won't finish and buffered on-chain config updates won't be applied. Note that the chain is still live in this case, but the validators cannot advance to new epochs.  
+
+*Mitigation*: Validators can force epoch change and disable randomness feature. The Aptos Labs need to deploy hotfix for full mitigation after debugging.
+
+**Randomness generation loses liveness**. If this happens, the blockchain will halt as every block needs randomness to proceed for execution.
+
+*Mitigation*: Validators can disable randomness feature using local configuration overwrite and restart all validators to pause the randomness generation to bring the chain back alive as a temporary mitigation. The Aptos Labs need to deploy hotfix for full mitigation after debugging. 
```

Comment:
@zjma add more details if necessary

Reply:
only implementation details are remaining. and i think they don't need to be in the AIP.